src/BotPlutusInterface/CoinSelection.hs

```diff
@@ -0,0 +1,497 @@
+        _ -> (idx :) <$> loop newTxInsVec' idxs vecs
+    loop _newTxInsVec [] [] = pure mempty
+    loop _newTxInsVec _idxs _vecs = Left "Lengths of indexes and list of vectors are not same."
+
+-- calculate euclidean distance of two vectors, of same length/dimension.
+l2norm :: ValueVector -> ValueVector -> Either Text Float
+l2norm v1 v2
+  | length v1 == length v2 = Right $ sqrt $ fromInteger $ sum $ Vec.zipWith formula v1 v2
+  | otherwise =
+    Left $
+      pack $
+        "Error: The length of the vectors should be same for l2norm. "
+          <> "length of vector v1: "
+          <> show (length v1)
+          <> " "
+          <> "length of vector v2: "
+          <> show (length v2)
+          <> "."
+  where
+    formula :: Integer -> Integer -> Integer
+    formula n1 n2 = (n1 - n2) ^ (2 :: Integer)
+
+-- Add two vectors of same length.
+addVec :: forall (n :: Type). Num n => Vector n -> Vector n -> Either Text (Vector n)
+addVec = opVec (+)
+
+-- Substract two vectors of same length.
+subVec :: forall (n :: Type). Num n => Vector n -> Vector n -> Either Text (Vector n)
+subVec = opVec (-)
+
+-- create zero vector of specified length.
+zeroVec :: Int -> Vector Integer
+zeroVec n = Vec.replicate n 0
+
+-- | Convert a value to a vector.
+valueToVec :: Set AssetClass -> Value -> Either Text ValueVector
+valueToVec allAssetClasses v =
+  maybeToRight "Error: Not able to uncons from empty vector." $
+    (over _Just fst . uncons) $ valuesToVecs allAssetClasses [v]
+
+-- | Convert values to a list of vectors.
+valuesToVecs :: Set AssetClass -> [Value] -> Vector ValueVector
+valuesToVecs allAssetClasses values = Vec.fromList $ map toVec values
+  where
+    toVec :: Value -> Vector Integer
+    toVec v =
+      fmap (Value.assetClassValueOf v) $
+        allAssetClasses & id %~ (Vec.fromList . Set.toList)
+
+-- | As the name suggests, we get a set of all the unique AssetClass from given the lists of values.
+uniqueAssetClasses :: [Value] -> Set AssetClass
+uniqueAssetClasses = Set.fromList . concatMap valueToAssetClass
+  where
+    valueToAssetClass :: Value -> [AssetClass]
+    valueToAssetClass = map (\(cs, tn, _) -> Value.assetClass cs tn) . Value.flattenValue
+
+-- Converting a chain index transaction output to a transaction input type
+txOutToTxIn :: (TxOutRef, TxOut) -> Either Text TxIn
+txOutToTxIn (txOutRef, txOut) =
+  case Ledger.addressCredential (txOutAddress txOut) of
+    PubKeyCredential _ -> Right $ Ledger.pubKeyTxIn txOutRef
+    ScriptCredential _ -> Left "Cannot covert a script output to TxIn"
+
+-- Apply a binary operation on two vectors of same length.
+opVec ::
+  forall (n :: Type).
+  Num n =>
+  (forall a. Num a => a -> a -> a) ->
+  Vector n ->
+  Vector n ->
+  Either Text (Vector n)
+opVec f v1 v2
+  | length v1 == length v2 = Right $ Vec.zipWith f v1 v2
+  | otherwise =
+    Left $
+      pack $
+        "Error: The length of the vectors should be same for addition."
```

Comment:
NIT: The error here speaks about addition, but this is also used for subVec, maybe we could phrase this in a generic way and also use it for l2norm

Reply:
Fixed it, thanks for pointing it out.